### PR TITLE
Add sidecar based maven and gradle Java stacks

### PR DESCRIPTION
**Pull request by l0rd**
### What does this PR do?

- Add stack java-maven
- Add stack java-gradle
- Modify samples to work with the new stacks
- Add a new spring-boot sample using this [spring guide repo](https://github.com/spring-guides/gs-rest-service)

**Some choices made to build the new stacks**:

- `che-theia:next` is used instead of `latest`
- I am using the default DockerHub images as runtime with JDK11 on Debian (i.e. [maven](https://hub.docker.com/_/maven) and [gradle](https://hub.docker.com/_/gradle))
- The runtime images tag is specified (don't use latest). I wanted to use sha256 tags (more reliable) but stacks API complained about that. Will create an issue.
- Memory:
  - che-theia: 512MB
  - maven runtime: 512MB
  - jdt.ls: 1GB
- The `${current.project.path}` placeholder for commands doesn't work on Theia and I have replaced it with `${CHE_PROJECTS_ROOT}/<sample-name>` 

**Current Limitations**
- JDT LS loading issue (related to https://github.com/theia-ide/theia/issues/3970)
- Java Debugging is not working (we need multiple vscode ext on the same sidecar #12395)
- JDT LS files are not persisted in a volume (that's related to meta.yaml and che-plugin.yaml issue to be created)

This PR is related to https://github.com/che-samples/console-java-simple/pull/7

@rhopp with that PR the good old che 6 java stack is gone. I suspect this will have some side effects on QE...

### CODE REVIEW

**Review by ghost — APPROVED** [review]

**Review by ashumilova — APPROVED** [review]

**Comment by benoitf** [comment]
`che-theia:next` no

**Review by benoitf — APPROVED** [review]

**Review by garagatyi — APPROVED** [review]

**Review by sleshchenko — APPROVED** [review]
LGTM

Please consider applying my proposals just to have tags in the same style

**Comment by ibuziuk** [comment]
I guess the plan is also to replace it for rh-che, right ?

**Comment by garagatyi** [comment]
Trying to run java extension in a sidecar on current master and get next error:
image
Does it work for somebody else?

**Comment by slemeur** [comment]
Nice.

I also saw a companion PR to update an existing sample: https://github.com/che-samples/console-java-simple
I think that can potentially break the sample in older version of Che?

**Comment by garagatyi** [comment]
I restarted workspace with Java plugin several times, changed default RAM, but still no java assistance. Is it known bug? Do we want to replace stack with such a state?

**Comment by l0rd** [comment]
> Nice.
> 
> I also saw a companion PR to update an existing sample: https://github.com/che-samples/console-java-simple
> I think that can potentially break the sample in older version of Che?

@slemeur it may break versions of Che that use a version of JDK older than JDK 8 (2014). Is that an issue?

> I restarted workspace with Java plugin several times, changed default RAM, but still no java assistance. Is it known bug? Do we want to replace stack with such a state?

@garagatyi agree that's annoying. I couldn't make it work neither. That's not related to JDT LS but to any vs code extension running in a sidecar. I have been pointed to https://github.com/theia-ide/theia/issues/3970. I was hoping that it could be fixed before this PR got merged because anyway it's critical. @benoitf @tolusha any comments about that?

That said if that cannot be fixed by the end of the sprint I am ok to not replace this stack but just create a distinct one (i.e. `java-maven`). But I hope we can proceed we the replacement and start moving forward with Che 7 (beta i.e. not perfect) stacks only.

By the way this make me think that even if we fix the JDT LS loading problem we still have the following limitations (added that to the issue description):
1. Java Debugging is not working (we need multiple vscode ext on the same sidecar)
2. JDT LS files are not persisted in a volume (that's related to )

**Comment by garagatyi** [comment]
@l0rd we will be able to create plugin that includes both java and java debugger soon https://github.com/eclipse/che/pull/12903

**Comment by benoitf** [comment]
@l0rd about https://github.com/theia-ide/theia/issues/3970 / plug-ins reload I'm working on it since the last couple of days and made several fixes. Another fix is pending https://github.com/theia-ide/theia/pull/4583 
Hopefully it may be the last one.

**Comment by l0rd** [comment]
@garagatyi @benoitf cool thank you for the update on those issues. So it looks like the situation is under control. @benoitf it would be great if you could test this java stack with your patch (see if JDT LS work fine) or if you could point me to a che-theia to use to test.

My main concern with this PR is QE team anyway. @rhopp I would like to merge it as it is to turn the page and don't think about Che 6 java stack anymore :-) But if that has a huge impact on your test I can keep (maybe hiding it) the existing default java stack.

**Comment by benoitf** [comment]
@l0rd I had issues on 'plug-in reload' with vscode-java extension and fixed it upstream https://github.com/redhat-developer/vscode-java/pull/832

**Comment by slemeur** [comment]
> @slemeur it may break versions of Che that use a version of JDK older than JDK 8 (2014). Is that an issue?
No, it's not 👍 Thanks for clarifying!

**Comment by SkorikSergey** [comment]
ci-test

**Comment by SkorikSergey** [comment]
Selenium tests execution on Eclipse Che Multiuser on OCP (https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/1634//Selenium_20tests_20report/) showed regression in selenium tests from ```dashboard``` package that use ```Java``` stack. 

Failed selenium tests will be updated by #12913 after this PR is merged.

File: ide/che-core-ide-templates/src/main/resources/samples.json [hunk 1]
```diff
@@ -989,6 +1044,69 @@
       "spring boot"
     ]
   },
+  {
+    "name": "spring-boot-gs-rest-service",
+    "displayName": "spring-boot-gs-rest-service",
+    "path": "/gs-rest-service",
+    "description": "Spring Boot Guide to build a RESTful web service ",
```
Comment (SkorikSergey) [inline ide/che-core-ide-templates/src/main/resources/samples.json]: Can you, please, remove space character in the end of ```spring-boot-gs-rest-service``` project description? Our selenium tests have some problem with it.

**Comment by l0rd** [comment]
For the record I will update this PR to keep the old GWT based java-default stack. I updating it because JDT LS is never loaded when using this stack hence there is no Java support yet (it should be related to #12904).

**Comment by l0rd** [comment]
@SkorikSergey I have updated the PR to keep the old java-default stack. We cannot replace the old stack until we have proper debugging support with JDT LS running in a sidecar (#12446) and I would like to merge this PR in the meantime. Is it ok for you to merge?

**Comment by SkorikSergey** [comment]
@l0rd It's Ok for me. There will need to make small changes to selenium tests. I will do it after merging.

**Comment by sunix** [comment]
@l0rd I am coming a bit late on this one:
Why don't we use the same container to host VSCode Java and java/graddle/maven command line tool ?
It makes sense that they all share the same versions. We would avoid duplication IMHO and would consume less memory.

**Comment by l0rd** [comment]
@sunix there is a thread in che-dev ([Custom builds and side cars](https://www.eclipse.org/lists/che-dev/msg03126.html)) where we discuss that. You may want to have a look at it and participate to the conversation?

**Pull request merged**